### ISSUE TITLE
Modal Preferences Window

### ISSUE THREAD

**Issue by daryl-williams** [tkambler/electron-preferences#77]
Hi Everyone, I am wondering if there is someway to make the preferences modal in nature, i.e. so it blocks until the user is done.

Tanks,

Daryl

**Comment by pvrobays** [comment]
Hi @daryl-williams 

That should definitely be possible as electron supports this.
I think this is what you'll need: [Browser Window - Modal windows](https://www.electronjs.org/docs/api/browser-window#modal-windows).

In essence: you'll have to pass the `parent: yourParentWindow` and  `modal: true` options to the BrowserOptions of the preference window. This can be done via the preferences json that's passed in the constructor:
```
const preferences = new ElectronPreferences({
    // ...
    'browserWindowOverrides': {
        'parent': yourParentWindow,
        'modal': true
    }
    // ...
});
```

Haven't had the time to try this out myself, but please let me know if this solution works for you.

For more info on the different options you can pass in the constructor, definitely check out the example application we have in the repo.

Pieter-Jan

**Comment by pvrobays** [comment]
Meanwhile I've been thinking about the fact that we're passing a complex object (BrowserWindow) to the options. We also pass this object over IPC to the application. However, electron doesn't like passing those around over IPC and their serialization of those objects fail.

So did a quick test and indeed: trying to make the preference window modal, crashes the app.
I've made a fix for this and created PR #79 

Meanwhile I've adapted the example app real quick to make the preference window modal. I'll drop the code here below for your reference. Allthough I wouldn't just copy paste this code as it's quite dirty (as is most of the example app) 😅

main.js
```
'use strict';

const electron = require('electron');
const app = electron.app;
const nativeTheme = electron.nativeTheme;
const BrowserWindow = electron.BrowserWindow;
const path = require('path');
const url = require('url');
const createPreferences = require('./preferences');

let mainWindow;

function createWindow() {
    
    mainWindow = new BrowserWindow({
        'width': 1200,
        'height': 700,
        'accept-first-mouse': true,
        webPreferences: {
            preload: path.join(__dirname, './Preload.js'),
            contextIsolation: true
        }
    });

    mainWindow.loadURL(url.format({
        'pathname': path.join(__dirname, 'index.html'),
        'protocol': 'file:',
        'slashes': true
    }));

    // mainWindow.webContents.openDevTools();

    mainWindow.on('closed', function() {
        mainWindow = null;
    });

    const preferences = createPreferences(mainWindow);
    
    nativeTheme.themeSource = preferences.preferences?.theme?.theme ?? "system";

    preferences.on('save', (preferences) => {
        console.log(`Preferences were saved.`, JSON.stringify(preferences, null, 4));

        nativeTheme.themeSource = preferences?.theme?.theme ?? "system";
    });
    
}

app.on('ready', createWindow);

app.on('window-all-closed', function() {
    if (process.platform !== 'darwin') {
        app.quit();
    }
});

app.on('activate', function() {
    if (mainWindow === null) {
        createWindow();
    }
});
```

preferences.js
```
'use strict';

// const electron = require('electron');
// const app = electron.app;
const path = require('path');
const os = require('os');
const ElectronPreferences = require('../');

const createPreferences = (parentWindow) => {
    return new ElectronPreferences({
        'dataStore': path.resolve(__dirname, 'preferences.json'),
        'defaults': {
            'notes': {
                'folder': path.resolve(os.homedir(), 'Notes')
            },
            'markdown': {
                'auto_format_links': true,
                'show_gutter': false
            },
            'preview': {
                'show': true
            },
            'drawer': {
                'show': true
            }
        },
        'webPreferences': {
            'devTools': true
        },
        'browserWindowOverrides': {
            'title': 'My Modal Electron Preferences',
            'parent': parentWindow,
            'modal': true
        },
        'sections': [
            {
                'id': 'about',
                'label': 'About You',
                'icon': 'single-01',
                'form': {
                    'groups': [
                        {
                            'label': 'About You',
                            'fields': [
                                {
                                    'label': 'First Name',
                                    'key': 'first_name',
                                    'type': 'text',
                                    'help': 'What is your first name?'
                                },
                                {
                                    'label': 'Last Name',
                                    'key': 'last_name',
                                    'type': 'text',
                                    'help': 'What is your last name?'
                                },
                                {
                                    'label': 'Gender',
                                    'key': 'gender',
                                    'type': 'dropdown',
                                    'options': [
                                        {'label': 'Male', 'value': 'male'},
                                        {'label': 'Female', 'value': 'female'},
                                        {'label': 'Unspecified', 'value': 'unspecified'},
                                    ],
                                    'help': 'What is your gender?'
                                },
                                {
                                    'label': 'Age',
                                    'key': 'age',
                                    'type': 'text',
                                    'inputType': 'number'
                                },
                                {
                                    'label': 'Which of the following foods do you like?',
                                    'key': 'foods',
                                    'type': 'checkbox',
                                    'options': [
                                        {'label': 'Ice Cream', 'value': 'ice_cream'},
                                        {'label': 'Carrots', 'value': 'carrots'},
                                        {'label': 'Cake', 'value': 'cake'},
                                        {'label': 'Spinach', 'value': 'spinach'}
                                    ],
                                    'help': 'Select one or more foods that you like.'
                                },
                                {
                                    'label': 'Coolness',
                                    'key': 'coolness',
                                    'type': 'slider',
                                    'min': 0,
                                    'max': 9001
                                },
                                {
                                    'label': 'Eye Color',
                                    'key': 'eye_color',
                                    'type': 'color',
                                    'format': 'hex',
                                    'help': 'Your eye color'
                                },
                                {
                                    'label': 'Hair Color',
                                    'key': 'hair_color',
                                    'type': 'color',
                                    'format': 'rgb',
                                    'help': 'Your hair color'
                                }
                            ]
                        }
                    ]
                }
            },
            {
                'id': 'notes',
                'label': 'Notes',
                'icon': 'folder-15',
                'form': {
                    'groups': [
                        {
                            'label': 'Stuff',
                            'fields': [
                                {
                                    'label': 'Read notes from folder',
                                    'key': 'folder',
                                    'type': 'directory',
                                    'help': 'The location where your notes will be stored.'
                                },
                                {
                                    'heading': 'Important Message',
                                    'content': '<p>The quick brown fox jumps over the long white fence. The quick brown fox jumps over the long white fence. The quick brown fox jumps over the long white fence. The quick brown fox jumps over the long white fence.</p>',
                                    'type': 'message',
                                }
                            ]
                        }
                    ]
                }
            },
            {
                'id': 'space',
                'label': 'Other Settings',
                'icon': 'spaceship',
                'form': {
                    'groups': [
                        {
                            'fields': [
                                {
                                    'label': 'Phone Number',
                                    'key': 'phone_number',
                                    'type': 'text',
                                    'help': 'What is your phone number?'
                                },
                                {
                                    'label': "Foo or Bar?",
                                    'key': 'foobar',
                                    'type': 'radio',
                                    'options': [
                                        {'label': 'Foo', 'value': 'foo'},
                                        {'label': 'Bar', 'value': 'bar'},
                                        {'label': 'FooBar', 'value': 'foobar'},
                                    ],
                                    'help': 'Foo? Bar?'
                                },
                                {
                                    'label': "Bar or Foo?",
                                    'key': 'barfoo',
                                    'type': 'radio',
                                    'options': [
                                        {'label': 'Bar', 'value': 'bar'},
                                        {'label': 'Foo', 'value': 'foo'},
                                        {'label': 'BarFoo', 'value': 'barfoo'},
                                    ],
                                    'help': 'Bar? Foo?'
                                },
                                {
                                    'label': 'Shortcut',
                                    'key': 'shortcut',
                                    'type': 'accelerator',
                                    'help': 'A keyboard shortcut'
                                }
                            ]
                        }
                    ]
                }
            },
            {
                'id': 'lists',
                'label': 'Lists',
                'icon': 'notes',
                'form': {
                    'groups': [
                        {
                            'label': 'Lists',
                            'fields': [
                                {
                                    'label': 'Favorite foods',
                                    'key': 'foods',
                                    'type': 'list',
                                    'size': 15,
                                    'help': 'A list of your favorite foods',
                                    'addItemValidator': /^[A-Za-z ]+$/.toString(),
                                    'addItemLabel': 'Add favorite food'
                                },
                                {
                                    'label': 'Best places to visit',
                                    'key': 'places',
                                    'type': 'list',
                                    'size': 10,
                                    'style': {
                                        'width': '75%'
                                    },
                                    'help': 'An ordered list of nice places to visit',
                                    'orderable': true
                                }
                            ]
                        }
                    ]
                }
            },
            {
                'id': 'theme',
                'label': 'Theme',
                'icon': 'brightness-6',
                'form': {
                    'groups': [
                        {
                            'fields': [
                                {
                                    'label': "Theme",
                                    'key': 'theme',
                                    'type': 'radio',
                                    'options': [
                                        {'label': 'System (default)', 'value': 'system'},
                                        {'label': 'Light', 'value': 'light'},
                                        {'label': 'Dark', 'value': 'dark'}
                                    ],
                                    'help': 'Light or dark theme?'
                                }
                            ]
                        }
                    ]
                }
            }
        ]
    });
}

module.exports = createPreferences;
```

**Comment by pvrobays** [comment]
The fix for this issue has been released in v2.3.2
